ssl/ssl_sess.c

```diff
@@ -832,31 +885,59 @@ int SSL_SESSION_set1_id(SSL_SESSION *s, const unsigned char *sid,
 
 long SSL_SESSION_set_timeout(SSL_SESSION *s, long t)
 {
-    if (s == NULL)
+    time_t new_timeout = (time_t)t;
+
+    if (s == NULL || t < 0)
         return 0;
-    s->timeout = t;
+    if (s->timeout != new_timeout) {
```

Comment:
Reads `s->timeout` not under lock, and then subsequently writes to it under lock. This seems wrong.

Reply:
Same as before.